crates/store/re_log_encoding/src/rrd/decoder/state_machine.rs

```diff
@@ -165,103 +212,184 @@ impl<T: DecoderEntrypoint> Decoder<T> {
                 let start_offset = self.byte_chunks.num_read() as u64;
 
                 if let Some(bytes) = self.byte_chunks.try_read(header.len as usize) {
-                    re_log::trace!(?header, "Read message");
-
                     let bytes_len = bytes.len() as u64;
                     let byte_span = re_chunk::Span {
                         start: start_offset,
                         len: bytes_len,
                     };
                     let message = match T::decode(
-                        bytes,
+                        bytes.clone(),
                         byte_span,
                         header.kind,
                         &mut self.app_id_cache,
                         self.version,
                     ) {
                         Ok(msg) => msg,
+
                         Err(err) => {
                             // We successfully parsed a header, but decided to drop the message altogether.
                             // We must go back to looking for headers, or the decoder will just be stuck in a dead
                             // state forever.
-                            self.state = DecoderState::MessageHeader;
+                            self.state = DecoderState::WaitingForMessageHeader;
                             return Err(err.into());
                         }
                     };
 
                     if let Some(message) = message {
-                        re_log::trace!("Decoded new message");
-
-                        self.state = DecoderState::MessageHeader;
+                        self.state = DecoderState::WaitingForMessageHeader;
                         return Ok(Some(message));
                     } else {
-                        re_log::trace!("End of stream - expecting a new Streamheader");
+                        re_log::trace!(
+                            "End of stream - expecting either a StreamFooter or a new Streamheader"
+                        );
+
+                        if !bytes.is_empty() {
+                            let rrd_footer =
+                                re_protos::log_msg::v1alpha1::RrdFooter::from_rrd_bytes(&bytes)?;
+                            _ = rrd_footer; // TODO(cmc): we'll use that in the next PR, promise.
+
+                            // A non-empty ::End message means there must be a footer ahead, no exception.
+                            self.state = DecoderState::WaitingForStreamFooter;
+                        } else {
+                            // There are 2 possible scenarios where we can end up here:
+                            // * The recording doesn't contain any data messages (e.g. it's empty except for
+                            //   a `SetStoreInfo` message).
+                            // * Backward compatibility: the payload is empty (i.e. `header.len == 0`) because the
+                            //   `End` message was written by a legacy encoder that predates the introduction of footers.
```

Comment:
That's bit of a spoiler: obviously as of this PR, the footer is still always empty.